desc/vmec_utils.py

```diff
@@ -469,3 +486,575 @@ def vmec_boundary_subspace(eq, RBC=None, ZBS=None, RBS=None, ZBC=None):  # noqa:
+
+    vol_grid = LinearGrid(M=2 * M_booz, N=2 * N_booz, NFP=eq.NFP, rho=r_half, sym=False)
+    # precompute the needed data for the boozer surface computations
+    # (except those which have to be computed on a single surface only,
+    # like theta_B, zeta_B or nu)
+    keys = [
+        "|B|",
+        "R",
+        "Z",
+        "sqrt(g)",
+        "rho",
+        "psi_r",
+        "lambda",
+        "B_zeta",
+        "B_theta",
+        "G",
+        "I",
+        "lambda_t",
+        "lambda_z",
+    ]
+    data_vol = eq.compute(keys, grid=vol_grid)
+
+    # FourierZernike fit the B_theta and B_zeta, then at each surface
+    # just evaluate the fit at the rho to get the B_theta_mn and B_zeta_mn
+    # so this we only only do a single fit
+    transform = Transform(
+        grid=vol_grid,
+        basis=FourierZernikeBasis(
+            L=2 * eq.L,
+            M=M_booz - 1,
+            N=N_booz,
+            sym="cos" if eq.sym else False,
+            NFP=eq.NFP,
+        ),
+        build_pinv=True,
+    )
+    B_zeta_lmn = transform.fit(data_vol["B_zeta"])
+    B_theta_lmn = transform.fit(data_vol["B_theta"])
+    # compute the fourier coeffs of the B_zeta and B_theta at each rho
+    _, _, B_zeta_mn = zernike_to_fourier(
+        B_zeta_lmn, transform.basis, rho=r_half, return_full=False
+    )
+    _, _, B_theta_mn = zernike_to_fourier(
+        B_theta_lmn, transform.basis, rho=r_half, return_full=False
+    )
+
+    B_transform = transforms["B"]
+    w_transform = transforms["w"]
+
+    data_keys = ["|B|_mn", "R_mn", "sqrt(g)_B_mn", "psi_r"]
+    data_keys = data_keys + ["Z_mn", "nu_mn"] if not eq.sym else data_keys
+    data_keys_sin = ["Z_mn", "nu_mn"]
+
+    @jit
+    def compute_data(grid, data):
+        trans = get_transforms(
+            ["|B|", "sqrt(g)"],
+            obj=eq,
+            grid=grid,
+            M_booz=M_booz - 1,
+            N_booz=N_booz,
+            jitable=True,
+        )
+
+        trans["B"] = B_transform
+        trans["w"] = w_transform
+
+        profiles = get_profiles(data_keys, obj=eq, grid=grid, jitable=True)
+        # compute data
+        d = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            data_keys,
+            params=eq.params_dict,
+            transforms=trans,
+            profiles=profiles,
+            data=data,
+        )
+
+        return d
+
+    @jit
+    def compute_data_sin_sym(grid, data):
+        # don't need to get transforms because all quantities that
+        # needed transforms other than the "B" and "w" are already
+        # computed from the compute_data call above
+        # and we have those transforms in transforms_sin
+        profiles = get_profiles(data_keys_sin, obj=eq, grid=grid, jitable=True)
+        # compute data
+        d = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            data_keys_sin,
+            params=eq.params_dict,
+            transforms=transforms_sin,
```

Comment:
Are you sure you don't have to compute fresh transforms here? Otherwise the grid could be different. 

If it is safe to use `transforms_sin` as pre-computed outside of this function, then why can't we do the same for `trans` in the above `compute_data` function? Or the `profiles` in either function?

Reply:
the `trans` in the `compute_data` function only is called so that we can get the correct theta_B, zeta_B for that rho surface (as those change from surface to surface, so we do have to re-create some transforms in the for loop to account for the fact that we need to eval things that have a rho dependence, that goes for why we need the get profiles too)

however the B_transform and w_transform are PURELY transforms that are a function of theta and zeta, so those only need to be built once and can be reused for every surface, hence why that is how we populate `trans["B"]` and `trans["w"]` in the comput_data function

now, when we get ot compute_data_sin, we have at this point already calculated `w` and `theta_B`, `zeta_B` etc, literally the only thing left that we must compute is `nu_mn` and `Z_mn`. However, these are sin-symmetric when stellarator sym is True, but the `trans["B"]` is cos symmetric by default when sym is True. Hence we need to use the `transforms_sin` where the transforms are sin symmetric to correctly caluclate `nu_mn` and `Z_mn`

The reason we don't need to re-create the transforms_sin at each step in the loop is because any rho-dependent data has already been computed with the `compute_data` call, so we can just reuse that, so the only transforms that matter in `compute_data_sin` are the transforms which only depend on the theta and zeta, which are precomputed before the loop

it is kind of confusing but this works and it agrees well when comparing computed quantities against the computation using the boozer modes, so I think it is correct according to the tests I have for it